### PR TITLE
issue 2608: Stop Playback of Recorder on last WebUI close

### LINKED ISSUE #2608: Recorder: Sound läuft weiter obwohl Tab geschlossen ist.
(kein Bug - aber führt evtl zu Überraschungen .)
Wenn der Recorder abspielt und man den Tab schliesst, bzw den kompletten Browser schliesst, läuft der Sound weiter.
Vielleicht kann man das zeitlich überwachen ?
Es kann ja ein langsames Überwachungsintervall sein. 2,3 oder sogar 5 Sekunden.

### PR DESCRIPTION
**Pull request by schlenkibus**
added overlay which will be prompted to the user under following conditions: the last recorder connection has been closed (closing the lid on an Laptop or exiting the browser or tab) and the recorder is replaying some audio. The prompt will ask the user whether or not he wants to stop the playback of the recorder, alternatively you can now stop the playback via the HWUI and WebUI Settings, closes #2608

### CODE REVIEW

File: projects/epc/audio-engine/src/recorder/Recorder.cpp [hunk 1]
```diff
@@ -187,3 +197,23 @@ nlohmann::json Recorder::queryFrames(FrameId begin, FrameId end) const
 
   return ret;
 }
+
+void Recorder::checkAndSendNoClientsStatus()
+{
+  while(true)
```
Comment (hhoegelo) [inline projects/epc/audio-engine/src/recorder/Recorder.cpp:203]: Huiuiui....
This thread never stops?
Anyway I don't see a reason for adding multi threading complexity and risk here. If you want a regular task to be performed, you could as well use an Expiration and refresh it, whenever the callback is invoked.

File: projects/epc/playground/src/proxies/hwui/panel-unit/boled/recorder/DoYouWantToStopRecorderLayout.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+{
+  if(down && i == Buttons::BUTTON_A)
+  {
+    Application::get().getHWUI()->getPanelUnit().getEditPanel().getBoled().resetOverlay();
+    return true;
+  }
+
+  if(down && i == Buttons::BUTTON_D)
+  {
+    RecorderManager::stopRecorder();
```
Comment (hhoegelo) [inline projects/epc/playground/src/proxies/hwui/panel-unit/boled/recorder/DoYouWantToStopRecorderLayout.cpp:35]: I think the class name, the Button class name, the message name and the method name in RecorderManager should reflect, that it is actually the 'playback' that is stopped. As a reader, I was under the impression that we stop the recording, if something is called 'stopRecorder'.

File: projects/epc/audio-engine/src/recorder/Recorder.h [hunk 3]
```diff
@@ -1,9 +1,11 @@
 #pragma once
 
 #include "Types.h"
+#include "nltools/threading/Throttler.h"
 #include <nltools/nlohmann/json.hpp>
 #include <memory>
 #include <sigc++/connection.h>
+#include <future>
```
Comment (hhoegelo) [inline projects/epc/audio-engine/src/recorder/Recorder.h:8]: can be removed

**Comment by schlenkibus** [comment]
@hhoegelo re-review pls

File: projects/epc/audio-engine/src/recorder/Recorder.h [hunk 4]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "Types.h"
+#include "nltools/threading/Throttler.h"
```
Comment (hhoegelo) [inline projects/epc/audio-engine/src/recorder/Recorder.h:4]: Throttler is not needed anymore

File: projects/epc/playground/src/use-cases/RecorderManager.cpp [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+sigc::connection RecorderManager::subscribeToNotifyNoRecorderUIsLeftAndStillPlaying(sigc::slot<void(void)> s)
+{
+  return m_signal.connect(s);
+}
+
+void RecorderManager::stopRecorderPlayback()
+{
+  nltools::msg::Setting::FlacRecorderStop msg {};
```
Comment (hhoegelo) [inline projects/epc/playground/src/use-cases/RecorderManager.cpp:23]: setting name should reflect, that playback is stopped, not recording

**Pull request merged**